File: src/core/validators/Association/AssociationMustNotDuplicateDomainEntityNames.cs
using MetaEd.Grammar.Antlr;

namespace MetaEd.Core.Validator.Association
{
    public class AssociationMustNotDuplicateDomainEntityNames : ValidationRuleBase<MetaEdGrammar.AssociationContext>
    {
        public override bool IsValid(MetaEdGrammar.AssociationContext context)
        {
            var firstDomainEntityName = context.firstDomainEntity().IdText();
            var secondDomainEntityName = context.secondDomainEntity().IdText();

            if (!firstDomainEntityName.Equals(secondDomainEntityName)) return true;

            // need to check 'with context' for naming collisions
            var firstContext = context.firstDomainEntity().withContext();
            var secondContext = context.secondDomainEntity().withContext();

            var firstContextName = firstContext == null ? string.Empty : firstContext.withContextName().ID().GetText();
            var secondContextName = secondContext == null ? string.Empty : secondContext.withContextName().ID().GetText();

            return !firstContextName.Equals(secondContextName);
        }

        public override string GetFailureMessage(MetaEdGrammar.AssociationContext context)
        {
            var identifier = context.associationName().GetText();
            var firstDomainEntityName = context.firstDomainEntity().IdText();

            return string.Format("Association '{0}' has duplicate declarations of Domain Entity '{1}'.", identifier, firstDomainEntityName);
        }
    }
}

File: src/core/validators/Association/AssociationMustNotDuplicateDomainEntityNames.ts
import {ValidationRuleBase} from "../ValidationRuleBase";
let MetaEdGrammar = require("../../../../src/grammar/gen/MetaEdGrammar").MetaEdGrammar;

export class AssociationMustNotDuplicateDomainEntityNames extends ValidationRuleBase
{
    public handlesContext(context: any) : boolean {
        return context.ruleIndex === MetaEdGrammar.RULE_association;
    }

    public isValid(context: any): boolean {
        let firstDomainEntityName = context.firstDomainEntity().propertyName().ID().getText();
        let secondDomainEntityName = context.secondDomainEntity().propertyName().ID().getText();
        if (firstDomainEntityName !== secondDomainEntityName)
            return true;
        let firstContext = context.firstDomainEntity().withContext();
        let secondContext = context.secondDomainEntity().withContext();
        let firstContextName = firstContext == null ? "" : firstContext.withContextName().ID().getText();
        let secondContextName = secondContext == null ? "" : secondContext.withContextName().ID().getText();
        return firstContextName !== secondContextName;
    }

    public getFailureMessage(context: any): string {
        let identifier = context.associationName().getText();
        let firstDomainEntityName = context.firstDomainEntity().propertyName().ID().getText();
        return `Association '${identifier}' has duplicate declarations of Domain Entity '${firstDomainEntityName}'`;
    }
}
